llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -188,54 +189,141 @@ int PostPipeliner::fit(MachineInstr *MI, int First, int Last, int II) {
+  computeForward();
+
+  // Backward properties like Latest and Offspring.
+  // Use a fixpoint loop, because plain reversed order may not be topological
+  // for predecessors
+  while (computeBackward()) {
+    /* EMPTY */;
+  }
+
+  // Adjust Earliest and Latest with resource requirements
+  for (int K = 0; K < NInstr; K++) {
+    auto &Me = Info[K];
+    SlotCounts ASlots;
+    for (int A : Me.Ancestors) {
+      auto &SU = DAG->SUnits[A];
+      ASlots += getSlotCounts(*SU.getInstr(), TII);
+    }
+    SlotCounts OSlots;
+    for (int O : Me.Offspring) {
+      auto &SU = DAG->SUnits[O];
+      OSlots += getSlotCounts(*SU.getInstr(), TII);
+    }
+    LLVM_DEBUG(dbgs() << "SU" << K << " : " << Info[K].Earliest << " - "
+                      << Info[K].Latest << " " << ASlots << " " << OSlots
+                      << "\n");
+    Me.Earliest = std::max(Me.Earliest, ASlots.max());
+    Me.Latest = std::min(Me.Latest, -OSlots.max());
+    LLVM_DEBUG(dbgs() << "    -> " << Info[K].Earliest << " - "
+                      << Info[K].Latest << "\n");
+  }
+
+  // Loop carried dependences will have pushed away Earliest of the second
+  // iteration, which should stay in lock step with the first.
   for (int K = 0; K < NInstr; K++) {
     const int K2 = K + NInstr;
     const int Earliest = Info[K2].Earliest - II;
     Info[K].Earliest = std::max(Info[K].Earliest, Earliest);
-    Info[K].Latest = Latest;
-  }
-
-  // Compute Latest. Use a fixpoint loop, because plain reversed
-  // order may not be topological for predecessors
-  bool Changed = true;
-  while (Changed) {
-    Changed = false;
-    for (int K = NInstr - 1; K >= 0; K--) {
-      SUnit &SU = DAG->SUnits[K];
-      const int Latest = Info[K].Latest;
-      for (auto &Dep : SU.Preds) {
-        int P = Dep.getSUnit()->NodeNum;
-        int NewLatest = Latest - Dep.getSignedLatency();
-        if (NewLatest < Info[P].Latest) {
-          Info[P].Latest = NewLatest;
-          Changed = true;
-        }
+  }
+
+  // Make Earliest of the second iteration push up Latest of the first
+  for (int K = 0; K < NInstr; K++) {
+    auto &Me = Info[K];
+    int LCDLatest = Me.Latest;
+    auto &SU = DAG->SUnits[K];
+    for (auto &Dep : SU.Succs) {
+      const int S = Dep.getSUnit()->NodeNum;
+      if (S < NInstr) {
+        continue;
       }
+      const int Earliest = Info[S - NInstr].Earliest;
+      const int Latest = Earliest - Dep.getSignedLatency();
+      LCDLatest = std::min(LCDLatest, Latest);
+    }
+    Me.LCDLatest = LCDLatest;
+    if (LCDLatest != Me.Latest) {
+      LLVM_DEBUG(dbgs() << "SU" << K << " LCDLatest=" << Me.LCDLatest << "\n");
     }
   }
 
-  LLVM_DEBUG(for (int K = 0; K < NInstr; K++) {
-    dbgs() << "SU" << K << " : " << Info[K].Earliest << " - " << Info[K].Latest
-           << "\n";
-  });
+  // Save the static values for ease of reset
+  for (auto &N : Info) {
+    N.StaticEarliest = N.Earliest;
+    N.StaticLatest = N.Latest;
```

Comment:
Instead of having `Latest`, `StaticLatest` and `LCDLatests`, could we only have `Latest` and `StaticLatest`? If some strategies need LCDLatest, I guess they could just call something like `biasLatestWithLCDs()`?

Reply:
Hmm. That might work now, but that is because we don't propagate LCDLatest backwards(yet), which makes it a local, relatively cheap computation. The situation is asymmetric relative to Earliest, where we do propagate it in Earliest itself. The root cause of this asymmetry is that we use the LCDs to constrain the scheduling intervals at the 'Earliest' side. If we would also constrain it at the Latest side, we constrain it twice, which might lead to suboptimality.
Summarizing, LCDLatest is a static heuristic value guiding TopDown scheduling, and might be difficult to compute if we want to propagate backwards. Latest is absolute and modified by the actual scheduling, and StaticLatest provides an expensive static component of Latest.

Reply:
Note that the cost to which 'relatively cheap'  applies here will be taken quite often, since it is computed for every comparison in the available queue.